agents/hwp_gripper/gripper_agent.py

```diff
@@ -0,0 +1,575 @@
+#!/usr/bin/env python3
+
+import argparse
+import ctypes
+import multiprocessing
+import os
+import signal
+import subprocess
+import sys
+import time
+
+import numpy as np
+
+this_dir = os.path.dirname(__file__)
+sys.path.append(
+    os.path.join(this_dir, 'src'))
+sys.path.append(
+    os.path.join(this_dir, 'pru'))
+
+import gripper_client as gclient
+import GripperBuilder as gb
+import GripperCollector as gc
+from ocs import ocs_agent, site_config
+from ocs.ocs_twisted import TimeoutLock
+
+# from socs.agents.hwp_supervisor.agent import get_op_data
+
+
+class GripperAgent:
+    def __init__(self, agent, mcu_ip, pru_port, control_port,
+                 return_port, supervisor_id=None, no_data_timeout=45 * 60):
+        self.agent = agent
+        self.log = agent.log
+        self.lock = TimeoutLock()
+
+        self._initialized = False
+        self.mcu_ip = mcu_ip
+        self.pru_port = pru_port
+        self.control_port = control_port
+        self.return_port = return_port
+
+        self.shutdown_mode = False
+        self.supervisor_id = supervisor_id
+        self.no_data_timeout = no_data_timeout
+
+        self.collector = None
+        self.builder = None
+        self.client = None
+
+        self.last_encoder = 0
+        self.last_limit = 0
+        self.last_limit_time = 0
+        self.is_forced = False
+
+        # self.encoder_names = ['Actuator 1 A', 'Actuator 1 B', 'Actuator 2 A',
+        #                      'Actuator 2 B', 'Actuator 3 A', 'Actuator 3 B']
+
+        self.encoder_pru = [0, 1, 2, 3, 4, 5]
+        self.encoder_edges = multiprocessing.Array(ctypes.c_int, (0, 0, 0, 0, 0, 0))
+        self.encoder_edges_record = multiprocessing.Array(ctypes.c_int, (0, 0, 0, 0, 0, 0))
+        self.encoder_direction = multiprocessing.Array(ctypes.c_int, (1, 1, 1, 1, 1, 1))
+
+        self.limit_names = ['Actuator 1 Cold', 'Actuator 1 Warm', 'Actuator 2 Cold',
+                            'Actuator 2 Warm', 'Actuator 3 Cold', 'Actuator 3 Warm']
+        self.limit_pru = [8, 9, 10, 11, 12, 13]
+        self.limit_state = [0, 0, 0, 0, 0, 0]
+
+        self.mode = multiprocessing.Value(ctypes.c_bool, False)
+        self.force = multiprocessing.Value(ctypes.c_bool, False)
```

Comment:
I've added some comments within the code which hopefully describes their function a bit better

Reply:
That was helpful, thanks. I have more comments in a re-review coming, but what I was working towards in trying to understand these attributes is that I think they should also be moved out of the Agent.

There's a lot of details with how the gripper data is interpreted that I don't think makes sense to have in the Agent class itself. All of these attributes seem to be required to ultimately call `self._get_pos()` in the `grip_acq` Process to determine the gripper position. This seems like it would fit nicely in another object in the driver code, perhaps one of the existing objects -- if not on the beaglebone end of things (though you should comment on whether this is feasible.)

I bring up that last point because I wonder what happens with this tracking if the Agent is restarted independent of any hardware power cycle of the gripper. The Agent is tracking all of these edge counts and converting them to position, but if the Agent is restarted we lose all that tracking. Does this mean we've lost track of the absolute position of the grippers? What do we do then? "Home" recalibrates the position, but can you always safely perform that operation whether gripped/ungripped/spinning/cold/warm?

Maybe that just moves the problem of persistence with these tracking variables to the beaglebone -- so what if that is restarted -- but I think it's more likely the Agent gets restarted than the beaglebone.